Refresh the buildVM script and use Pyrex

amstewart opened the pull request:
The AZDO pipeline build machines do not natively include ISO and VM utilities like `genisoimage`, `qemu-utils`, and the QEMU system binaries. So they cannot natively run the `buildVM.sh` script. Since they are already using the Pyrex container, we may as well just add these utilities to that container context.

And since I already need to touch the `buildVM` script to make that work, I decided to just make it a first class scriptlet in the `scripts/pipelines` directory. This PR moves the `buildVM` script into a `build.vms.sh` scriptlet, and majorly refreshes it to be simpler and more modern. Most importantly, it now has sane defaults. Simply calling the script from anywhere in the repo, after the recovery media has been created, is sufficient to create a QEMU VM archive.

Otherwise, the major steps of the build-vm script haven't fundamentally changed.

# Testing
Tested many times on my dev machine.

@ni/rtos

amstewart committed:
scripts/pipelines: add build.vm.sh

Create a build.vm pipeline script based on the buildVM.sh script.

Signed-off-by: Alex Stewart <alex.stewart@ni.com>

amstewart committed:
fixup! scripts/pipelines: add build.vm.sh

amstewart committed:
add qemu and iso utilities to Pyrex container

Not all NILRT build machines will necessarily have ISO and VM utilities
installed to them. Since the official VM creation pipeline requires
them, we may as well include them with the Pyrex container, and use its
environment to create the VMs.

Signed-off-by: Alex Stewart <alex.stewart@ni.com>

amstewart committed:
pyrex: enable KVM passthrough

Pass through the /dev/kvm device from the docker host, so that it can be
used by the pyrex container when running QEMU.

Signed-off-by: Alex Stewart <alex.stewart@ni.com>

amstewart committed:
build.vm.sh: remove extraneous image_dir argument

The image_dir argument was only used to define a path to the
recovery_iso, which can already be independently overwritten. Replace it
instead with a default value that will only apply in the case that the
user hasn't specified a recovery ISO.

Signed-off-by: Alex Stewart <alex.stewart@ni.com>

amstewart committed:
scripts: move buildVM-files

The buildVM script has been replaced by the build.vm.sh pipeline script.
Move the vm resources directory into the scripts/pipelines/ directory,
so that it is co-located with the new script.

Signed-off-by: Alex Stewart <alex.stewart@ni.com>

amstewart committed:
docker: better documentation for Dockerfile

Signed-off-by: Alex Stewart <alex.stewart@ni.com>

amstewart commented:
@shruthi-ravi Let's hold off on merging this until I'm back from the holidays. I'm not 100% sure that the new Dockerfile won't break the AZDO pipeline until I push a new container manually.

shruthi-ravi commented on scripts/pipelines/build.vms.sh:
```diff
@@ -0,0 +1,261 @@
+
+# Source the common build setup script, so that we're pyrex-enabled and our
+# working directory is changed to the build directory.
+. "${SCRIPT_ROOT}/build.common.sh"
```
Does this work fine if this script is run in an already Pyrex-enabled environment? I haven't set up the Pyrex stuff yet, so I don't know.

amstewart replied:
Yes. If you look at the other files in `:scripts/pipelines/`, they all start by sourcing the `build.common.sh` script. Pyrex initialization doesn't drop the build machine into a persistent container context, it just sets up the build/ directory in a special way. So you can arbitrarily run through the "initialization" codepaths afterward, and they will still execute in the host machine context. It's only when you run one of the pyrex-container commands that the container is started.

shruthi-ravi commented on scripts/pipelines/build.vms.sh:
```diff
@@ -0,0 +1,261 @@
+
+# Environmentals
+MACHINE
+	The machine architecture of the created VM. Only the 'x64' architecture is
```
How do you control the value of this variable? i.e. restrict it to only x64? Why not have it as a constant for now? Then, we don't have to do any validation on the value.

amstewart replied:
I don't restrict it anywhere. We've used the `MACHINE` variable in so many NILRT scripts that I was thinking of it as a *standard* switch that NILRT builders will know how to use. And I continued to use it here in the case that we add more architecture support in the future.

But now that you have me thinking about it again - those aren't good reasons. If we add more arch-support in the future, we will need to refactor anyway. I'll drop the variability and hardcode the `x64` value for now, for simplicity.

shruthi-ravi commented on scripts/pipelines/build.vms.sh:
```diff
@@ -0,0 +1,261 @@
+	The desired size (in megabytes) for the VM's virtual RAM. Defaults to
+    1024 MB.
+-n,--name NAME
+    The desired name for the VM archive.
```
Defaults to "nilrt".

amstewart replied:
Will fix.

shruthi-ravi commented on scripts/pipelines/build.vms.sh:
```diff
@@ -0,0 +1,261 @@
+		"./${script_destination_path}"
+
+    sed -i "s%\${PRIMARY_DISK}%${primary_disk}%g" "${script_destination_path}"
+    sed -i "s%\${VM_MACHINE_UUID}%${vm_uuid}%g" "${script_destination_path}"
+    sed -i "s%\${VM_DISK_UUID}%${disk_uuid}%g" "${script_destination_path}"
```
These variables seem to be used only in the old .vbox templates.

amstewart replied:
True. I'll remove them.

amstewart committed:
build.vms.sh: remove unused UUID template vars

The VM_*_UUID template variables were only used by the now-deprecated
VMware machine definition templates. Remove them, since there are no
users.

Signed-off-by: Alex Stewart <alex.stewart@ni.com>

amstewart committed:
build.vms.sh: expose default name in usage text

Signed-off-by: Alex Stewart <alex.stewart@ni.com>

amstewart committed:
build.vms.sh: hardcode MACHINE to x64

NILRT currently only supports the x64 MACHINE value, and the build.vms
script only works with QEMU x86_64 emulation. There is no need to
complicate the build.vms script by pretending that we support more
architectures.

Remove any references to MACHINE or variable architecture and replace
them with hardcoded 'x64' values.

Signed-off-by: Alex Stewart <alex.stewart@ni.com>

amstewart commented:
# Patch V2
* Addressed @shruthi-ravi's feedback.
    * Removed unused UUID template vars.
    * Updated usage text.
    * Hardcoded the `x64` MACHINE value.
